### PR TITLE
Fix a race condition around some methods that get deduplicated

### PR DESCRIPTION
**Pull request by poljar**
The first patch is the meat of the PR, i suggest this to be reviewed commit by commit.

The second patch makes it easier to ensure that some fields, i.e. the `members_synced` field, are only modified in one place of the codebase.

The third patch makes sure that syncing members only happens if it isn't already done so.

### CODE REVIEW

File: crates/matrix-sdk/src/room/common.rs [hunk 1]
```diff
@@ -391,16 +391,16 @@ impl Common {
         }
     }
 
-    pub(crate) async fn ensure_members(&self) -> Result<()> {
+    pub(crate) async fn ensure_members(&self) -> Result<Option<MembersResponse>> {
```
Comment (Hywan) [inline crates/matrix-sdk/src/room/common.rs:394]: :+1:

File: crates/matrix-sdk/src/client/mod.rs [hunk 2]
```diff
@@ -151,11 +152,11 @@ pub(crate) struct ClientInner {
     /// Locks making sure we only have one group session sharing request in
     /// flight per room.
     #[cfg(feature = "e2e-encryption")]
-    pub(crate) group_session_locks: DashMap<OwnedRoomId, Arc<Mutex<()>>>,
+    pub(crate) group_session_locks: Mutex<BTreeMap<OwnedRoomId, Arc<Mutex<()>>>>,
     /// Lock making sure we're only doing one key claim request at a time.
     #[cfg(feature = "e2e-encryption")]
     pub(crate) key_claim_lock: Mutex<()>,
-    pub(crate) members_request_locks: DashMap<OwnedRoomId, Arc<Mutex<()>>>,
+    pub(crate) members_request_locks: Mutex<BTreeMap<OwnedRoomId, Arc<Mutex<()>>>>,
```
Comment (Hywan) [inline crates/matrix-sdk/src/client/mod.rs:159]: :+1:

File: crates/matrix-sdk/src/room/common.rs [hunk 3]
```diff
@@ -301,7 +301,7 @@ impl Common {
     pub(crate) async fn request_members(&self) -> Result<Option<MembersResponse>> {
         let mut map = self.client.inner.members_request_locks.lock().await;
 
-        if let Some(mutex) = map.get(self.inner.room_id()).map(|m| m.clone()) {
+        if let Some(mutex) = map.get(self.inner.room_id()).cloned() {
```
Comment (Hywan) [inline crates/matrix-sdk/src/room/common.rs:304]: Can you rebase to merge this commit with the first one please?

**Review by Hywan — APPROVED** [review]
Looks good to me, well done!

File: crates/matrix-sdk/src/room/common.rs (hunk 3 above)
Comment (poljar) [reply crates/matrix-sdk/src/room/common.rs:304]: Yup, I ~always autosquash the fixup commits after review.

**Pull request merged**